interfaces/builtin/fpga.go

```diff
@@ -0,0 +1,65 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+const fpgaSummary = `allows access to the FPGA subsystem`
+
+const fpgaBaseDeclarationSlots = `
+  fpga:
+    allow-installation:
+      slot-snap-type:
+        - core
+`
+
+const fpgaConnectedPlugAppArmor = `
+# Description: Can access fpga subsystem.
+
+# Devices
+/dev/fpga[0-9]* rw,
```

Comment:
this will also need a udev device cgroup rule too, do you have a system with these fpga devices? what is the output of 

```
udevadm info -n /dev/fpga0 -q property
```

Reply:
It is:
```
$ udevadm info -n /dev/fpga0 -q property
DEVPATH=/devices/virtual/misc/fpga0
DEVNAME=/dev/fpga0
MAJOR=10
MINOR=59
SUBSYSTEM=misc
```
Not sure if we need the udev plug here, is not that needed only if the device is not in sysfs too? This one is.

Reply:
For all apparmor accesses to `/dev/...` in an interface, we should always have associated udev rules, which put the snap into a device cgroup.

Currently, we have the problem that if a snap plugs interfaces that don't declare any udev rules, the snap is not put into a device cgroup, but we want to have "security in depth" as per @jdstrand's explanation somewhere such that access to devices is mediated by both apparmor _and_ the device cgroup. We eventually want to put all snaps into a device cgroup by default, however that was blocked by some problems and we haven't had a chance to revisit that yet. Because we don't currently put snaps into the device cgroup unless there are any interfaces declaring udev rules to add devices to the device cgroup, if you built a version of snapd with this interface and then tested it out with a snap that only declared this plug, you would find that this interface worked, but adding another interface which triggers the device cgroup transition such as network-control, or opengl, etc. then you would find your interface definition won't work without the udev rules.

Hopefully that all makes sense

Reply:
as to what udev rule you should use, I'm not sure actually given that this is in the misc subsystem, we may need a more complicated rule

Reply:
@anonymouse64 thanks for the detailed explanation, I had forgotten this thing about the cgroup being created when you add an interface that has udev connected plug. I have re-pushed adding udev rules.

Reply:
@anonymouse64 I am not as familiar with udev rules for cgroup setup in snapd - does the one proposed by @alfonsosanchezbeato seem reasonable in this case? From what I can see it does, but as you asked the intial question it would be good to get your ack on this.

Reply:
yes, the rules here look good to me, thanks for adding it